packages/protocol/contracts/stability/SortedOracles.sol

```diff
@@ -139,54 +137,64 @@ contract SortedOracles is ISortedOracles, Ownable, Initializable {
-   * @dev Values are passed as uint128 to avoid uint256 overflow when multiplying.
    * @dev Note that only one of `lesserKey` or `greaterKey` needs to be correct to reduce friction.
    */
   function report(
     address token,
-    uint128 numerator,
-    uint128 denominator,
+    uint256 numerator,
+    uint256 denominator,
     address lesserKey,
     address greaterKey
   )
     external
     onlyOracle(token)
   {
-    SortedFractionMedianList.Element memory originalMedian = getMedianElement(rates[token]);
-    rates[token].insertOrUpdate(msg.sender, numerator, denominator, lesserKey, greaterKey);
-    timestamps[token].insertOrUpdate(
-      msg.sender,
-      // solhint-disable-next-line not-rely-on-time
-      uint128(now),
-      1,
-      getLesserTimestampKey(token, msg.sender),
-      address(0)
-    );
-    // solhint-disable-next-line not-rely-on-time
-    emit OracleReported(token, msg.sender, now, numerator, denominator);
-    emitIfMedianUpdated(token, originalMedian);
+    uint256 originalMedian = rates[token].getMedianValue();
+    uint256 value = numerator.mul(DENOMINATOR).div(denominator);
+    if (rates[token].contains(msg.sender)) {
+      rates[token].update(msg.sender, value, lesserKey, greaterKey);
+      timestamps[token].update(
+        msg.sender,
+        // solhint-disable-next-line not-rely-on-time
+        now,
+        timestamps[token].getHead(),
+        address(0)
+      );
+    } else {
+      rates[token].insert(msg.sender, value, lesserKey, greaterKey);
+      timestamps[token].insert(
+        msg.sender,
+        // solhint-disable-next-line not-rely-on-time
+        now,
+        timestamps[token].getHead(),
+        address(0)
+      );
+    }
+    emit OracleReported(token, msg.sender, now, value, DENOMINATOR);
+    uint256 newMedian = rates[token].getMedianValue();
+    if (newMedian != originalMedian) {
+      emit MedianUpdated(token, newMedian, DENOMINATOR);
+    }
   }
 
   /**
    * @notice Returns the number of rates.
    * @param token The address of the token for which the Celo Gold exchange rate is being reported.
    * @return The number of reported oracle rates for `token`.
    */
-  function numRates(address token) external view returns (uint256) {
-    return rates[token].numElements;
+  function numRates(address token) public view returns (uint256) {
+    return rates[token].getNumElements();
   }
 
   /**
    * @notice Returns the median rate.
    * @param token The address of the token for which the Celo Gold exchange rate is being reported.
    * @return The median exchange rate for `token`.
    */
-  function medianRate(address token) external view returns (uint128, uint128) {
-    SortedFractionMedianList.Element memory median = getMedianElement(rates[token]);
-    return (median.numerator, median.denominator);
+  function medianRate(address token) external view returns (uint256, uint256) {
+    return (rates[token].getMedianValue(), numRates(token) == 0 ? 0 : DENOMINATOR);
```

Comment:
Will this lead to inadvertent divide by zero errors?

Reply:
It will.  It's supposed to be invalid for there to be zero oracle rates reported. Any suggestions for how to enforce this?

Reply:
Upon initialization insert 1 report and prevent removals which decrease number of elements below 1?